Prepare to make RenderObjectElement buildable

abarth opened the pull request:
This patch prepares us to pass a BuildContext to RenderObjectWidgets, which
will make it possible to rebuild RenderObjectElements:
- Delay creation of the render object until mount(). That will let us pass
  `this` to createRenderObject and have the inherited elements be initialized.
- Cleanup widgets that take builder closures to have consistent semantics and
  document those semantics in the Builder widget.

Related to #2598

abarth committed:
Prepare to make RenderObjectElement buildable

This patch prepares us to pass a BuildContext to RenderObjectWidgets, which
will make it possible to rebuild RenderObjectElements:

 * Delay creation of the render object until mount(). That will let us pass
   `this` to createRenderObject and have the inherited elements be initialized.

 * Cleanup widgets that take builder closures to prepare for their
   RenderObjectElement to be rebuilt more often.

 * Add a test for the interaction between inherited widgets and
   MixedViewport.

Related to #2598

abarth commented:
@Hixie

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1530,29 +1528,11 @@ abstract class RenderObjectElement<T extends RenderObjectWidget> extends Buildab
-    // implement reinvokeBuilders() if they do provide a way to mark themeselves
-    // dirty, e.g. if they have a builder callback. (Builder callbacks have a
-    // 'BuildContext' argument which you can pass to Theme.of() and other
-    // InheritedWidget APIs which eventually trigger a rebuild.)
```
I remember now. This comment is how you end up here. Suppose your MixedViewport has a builder function that does `Foo.of(context)`. The `context` here will be this Element.

abarth replied:
Ah, I'll add a test for that case.

Hixie commented:
If you add a test to prove that a builder function that uses an inherited .of() still works ok, LGTM. But I do think that's where we'll find issues with this.

abarth commented:
The test showed that there was an issue, but I've fixed it.

I've reverted the change to VirtualViewport because it caused a test to fail but I've added a TODO about virtual viewport having quirky semantics.

Hixie commented:
lgtm
